Sources/SwiftSyntax/SourceLocation.swift

```diff
@@ -0,0 +1,344 @@
+      data.position
+    return converter.location(for: pos)
+  }
+
+  /// The ending location, in the provided file, of this Syntax node.
+  /// - Parameters:
+  ///   - converter: The `SourceLocationConverter` that was previously
+  ///     initialized using the root tree of this node.
+  ///   - afterTrailingTrivia: Whether to skip trailing trivia when getting
+  ///                          the node's location. Defaults to `false`.
+  public func endLocation(
+    converter: SourceLocationConverter,
+    afterTrailingTrivia: Bool = false
+  ) -> SourceLocation {
+    var pos = data.position
+    pos += raw.leadingTriviaLength
+    pos += raw.contentLength
+    if afterTrailingTrivia {
+      pos += raw.trailingTriviaLength
+    }
+    return converter.location(for: pos)
+  }
+
+  /// The source range, in the provided file, of this Syntax node.
+  /// - Parameters:
+  ///   - converter: The `SourceLocationConverter` that was previously
+  ///     initialized using the root tree of this node.
+  ///   - afterLeadingTrivia: Whether to skip leading trivia when getting
+  ///                          the node's start location. Defaults to `true`.
+  ///   - afterTrailingTrivia: Whether to skip trailing trivia when getting
+  ///                          the node's end location. Defaults to `false`.
+  public func sourceRange(
+    converter: SourceLocationConverter,
+    afterLeadingTrivia: Bool = true,
+    afterTrailingTrivia: Bool = false
+  ) -> SourceRange {
+    let start = startLocation(converter: converter, afterLeadingTrivia: afterLeadingTrivia)
+    let end = endLocation(converter: converter, afterTrailingTrivia: afterTrailingTrivia)
+    return SourceRange(start: start, end: end)
+  }
+}
+
+/// Returns array of lines with the position at the start of the line and
+/// the end-of-file position.
+fileprivate func computeLines(
+  tree: SourceFileSyntax
+) -> ([AbsolutePosition], AbsolutePosition) {
+  var lines: [AbsolutePosition] = []
+  // First line starts from the beginning.
+  lines.append(.startOfFile)
+  var position: AbsolutePosition = .startOfFile
+  let addLine = { (lineLength: SourceLength) in
+    position += lineLength
+    lines.append(position)
+  }
+  var curPrefix: SourceLength = .zero
+  tree.forEachToken {
+    curPrefix = $0.forEachLineLength(prefix: curPrefix, body: addLine)
+  }
+  position += curPrefix
+  return (lines, position)
+}
+
+fileprivate extension String {
+  /// Walks and passes to `body` the `SourceLength` for every detected line,
+  /// with the newline character included.
+  /// - Returns: The leftover `SourceLength` at the end of the walk.
+  func forEachLineLength(
+    prefix: SourceLength = .zero, body: (SourceLength) -> ()
+  ) -> SourceLength {
+    var lineLength = prefix
+    for char in self {
+      lineLength += SourceLength(utf8Length: String(char).utf8.count)
+      if char.isNewline {
+        body(lineLength)
+        lineLength = .zero
+      }
```

Comment:
I really don't like this block... How about:
```swift
for i in self.indices {
  lineLength += SourceLength(utf8Length: self[i...i].utf8.count)
  if (self[i].isNewline) {
    body(lineLength)
    lineLength = .zero
  }
}
```
I think slicing is faster than `String(char)`.

Reply:
`self[i...i]` may search character boundary (again). So
```swift
var i = self.startIndex
while i != self.endIndex {
  var next = self.index(after: i)
  lineLength += SourceLength(utf8Length: self[i..<next].utf8.count)
  if (self[i].isNewline) {
    body(lineLength)
    lineLength = .zero
  }
  i = next
}
```
might be faster.

Reply:
I've asked @milseman and he believed `String(char)` is not a performance concern since it will be a small string. Michael could you verify if there is a performance difference between `String(char)` and Rintaro's slicing alternative ?

Reply:
`String(char).utf8.count` should be the fastest way to get this information until we make [this](https://github.com/apple/swift/blob/master/stdlib/public/core/Character.swift#L93) public (right after Swift 5.0).

Another question, what do you want the semantics of this method to be? Is your notion of newline the same as Character's? Also, should SwiftSyntax's processing be post-grapheme-segmentation?

Reply:
> Another question, what do you want the semantics of this method to be? Is your notion of newline the same as Character's?

I feel that this is a sensible default. Clients could do some input sanitization if they want more control. For example an editor using swift/sourcekitd needs to sanitize new lines to use ascii new-line because that's the only thing swift/sourcekitd understand as a new-line.

> Also, should SwiftSyntax's processing be post-grapheme-segmentation?

I don't understand the question, could you explain more ?

Reply:
For example, "\n\u{301}" is technical a single character, it's a newline with a combining accent printed over it. But, you may want to view this as a newline followed by a combining accent scalar, which is what you'll get from `unicodeScalars` instead. Also, do you want canonical equivalence for comparisons, or to compare the exact sequence of scalars?

Reply:
> Is your notion of newline the same as Character’s?

The [Swift grammar](https://docs.swift.org/swift-book/ReferenceManual/LexicalStructure.html#grammar_line-break) specifies otherwise.

Reply:
So maybe this is best served as:

```swift
for char in self {
  lineLength += SourceLength(utf8Length: String(char).utf8.count)
  if char == "\n" || char == "\r" || char == "\r\n" {
    body(lineLength)
    lineLength = .zero
  }
}
```

Reply:
@SDGGiesbrecht thanks for pointing to the grammar document! I find it disconcerting that the swift compiler actually does not consider a single `'\r'` as a new line, it only considers `'\n'`.

@milseman what do you think is more performant, doing:

```
switch char {
case "\n", "\r\n", "\r":
   <code>
default:
    break
}
```
or
```
if let val = char.asciiValue, val == 10 || val == 13 {
  <code>
}
```
?
Note that `char.asciiValue` automatically converts `"\r\n"` to `'\n'`

Reply:
@milseman mentioned the fastest would be to iterate over the `UTF8View` and check the bytes, I'll make such a change.